### DISCUSSION TITLE
Prisma + Cloud Run + Cloud SQL + Github Actions

### DISCUSSION THREAD

**Discussion by tomwray13** [prisma/prisma#13006] [Q&A]
Hey there!

I've got a simple NestJS project that I've been playing around with to learn how to use Github Actions to setup a simple CD workflow for deploying to Cloud Run.

I've got the above all working - the next thing I'm trying to do is add Prisma for the database layer and use Cloud SQL as the hosted instance for production.

What I'm struggling with is **where** and **when** to run the `npx prisma migrate deploy` command so the database migrations are applied to the Cloud SQL datbase instance. There is some documentation on the Prisma website for Heroku, but not for Google Cloud Platforms Cloud Run + Cloud SQL combo.

What I'm doing currently is running the `npx prisma migrate deploy` command as the final step of the Github Action, however, I'm getting some errors back. Here's the github action:

```
name: Deploy to Cloud Run from Source

on:
  push:
    branches:
      - main

env:
  PROJECT_ID: direct-album-348214 # TODO: update Google Cloud project id
  SERVICE: nest-fun # TODO: update Cloud Run service name
  REGION: us-west1 # TODO: update Cloud Run service region
  DATABASE_URL: postgresql://tom:secret@localhost/postgres?host=/cloudsql/direct-album-348214:us-central1:nest-fun-postgres

jobs:
  deploy:
    # Add 'id-token' with the intended permissions for workload identity federation
    permissions:
      contents: 'read'
      id-token: 'write'

    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Google Auth
        id: auth
        uses: 'google-github-actions/auth@v0'
        with:
          workload_identity_provider: '${{ secrets.WIF_PROVIDER }}' # e.g. - projects/123456789/locations/global/workloadIdentityPools/my-pool/providers/my-provider
          service_account: '${{ secrets.WIF_SERVICE_ACCOUNT }}' # e.g. - my-service-account@my-project.iam.gserviceaccount.com

      - name: Deploy to Cloud Run
        id: deploy
        uses: google-github-actions/deploy-cloudrun@v0
        with:
          service: ${{ env.SERVICE }}
          region: ${{ env.REGION }}
          # NOTE: If required, update to the appropriate source folder
          source: ./
      
      - name: Run database migrations
        run: "npx prisma migrate deploy"
        

      # If required, use the Cloud Run url output in later steps
      - name: Show Output
        run: echo ${{ steps.deploy.outputs.url }}

```

The Github action works until the **Run database migrations** step where it then spits out this error:

```
npx prisma migrate deploy
  shell: /usr/bin/bash -e {0}
  env:
    PROJECT_ID: direct-album-[3](https://github.com/tomwray13/nest-fun/runs/6172901912?check_suite_focus=true#step:5:3)[4](https://github.com/tomwray13/nest-fun/runs/6172901912?check_suite_focus=true#step:5:4)8214
    SERVICE: nest-fun
    REGION: us-west1
    DATABASE_URL: ***localhost/postgres?host=/cloudsql/direct-album-348214:us-central1:nest-fun-postgres
    CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE: /home/runner/work/nest-fun/nest-fun/gha-creds-01f1e2be[5](https://github.com/tomwray13/nest-fun/runs/6172901912?check_suite_focus=true#step:5:5)f0294ea.json
    GOOGLE_APPLICATION_CREDENTIALS: /home/runner/work/nest-fun/nest-fun/gha-creds-01f1e2be5f0294ea.json
    GOOGLE_GHA_CREDS_PATH: /home/runner/work/nest-fun/nest-fun/gha-creds-01f1e2be5f0294ea.json
    CLOUDSDK_CORE_PROJECT: direct-album-34[8](https://github.com/tomwray13/nest-fun/runs/6172901912?check_suite_focus=true#step:5:8)214
    CLOUDSDK_PROJECT: direct-album-348214
    GCLOUD_PROJECT: direct-album-348214
    GCP_PROJECT: direct-album-348214
    GOOGLE_CLOUD_PROJECT: direct-album-348214
    CLOUDSDK_METRICS_ENVIRONMENT: github-actions-deploy-cloudrun
npm WARN exec The following package was not found and will be installed: prisma
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "postgres", schema "public" at "localhost:5432"

Error: P[10](https://github.com/tomwray13/nest-fun/runs/6172901912?check_suite_focus=true#step:5:10)01: Can't reach database server at `/cloudsql/direct-album-3482[14](https://github.com/tomwray13/nest-fun/runs/6172901912?check_suite_focus=true#step:5:14):us-central1:nest-fun-postgres`:`5432`

Please make sure your database server is running at `/cloudsql/direct-album-348[21](https://github.com/tomwray13/nest-fun/runs/6172901912?check_suite_focus=true#step:5:21)4:us-central1:nest-fun-postgres`:`5432`.
Error: Process completed with exit code 1.
```

It feels like there should be a step in the Github Action that opens up a terminal against the Cloud Run container, and then run the "npx prisma migrate deploy" command against that container, using the DATABASE_URL value that would be set as an env variable in the Cloud Run instance - but I'm not sure if this is correct.

Any help on the above would be hugely appreciated!

P.S This is just a testing project - I'm aware the above code exposes some sensitive info like the database creds.

**Comment by austincrim** [comment]
Hey @tomwray13 👋 ,

I'm not very familiar with Cloud SQL, so I have a couple questions.

- Is the Cloud SQL instance accessible via the public internet?
- Is it expected that the connection string contains `localhost`?
- Can you connect to the Cloud SQL instance when developing locally?

**Reply by tomwray13** [reply]
Hey @austincrim! Thanks for your questions and help. 

I've managed to figure it out - what I needed to do what connect to the Cloud SQL Proxy within the Github Action and then run the prisma migration after. For anyone else with this issue in the future, here's how my Github Action file looks now with the fixes (note the final few steps which I took from the [GCP Cloud SQL Proxy documentation](https://cloud.google.com/sql/docs/postgres/connect-admin-proxy)):

```
name: Deploy to Cloud Run from Source

on:
  push:
    branches:
      - main

env:
  PROJECT_ID: direct-album-348214
  SERVICE: nest-fun
  REGION: us-west1
  DATABASE_URL: postgresql://tom:secret@localhost:5432/postgres

jobs:
  deploy:
    # Add 'id-token' with the intended permissions for workload identity federation
    permissions:
      contents: 'read'
      id-token: 'write'

    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Google Auth
        id: auth
        uses: 'google-github-actions/auth@v0'
        with:
          workload_identity_provider: '${{ secrets.WIF_PROVIDER }}' # e.g. - projects/123456789/locations/global/workloadIdentityPools/my-pool/providers/my-provider
          service_account: '${{ secrets.WIF_SERVICE_ACCOUNT }}' # e.g. - my-service-account@my-project.iam.gserviceaccount.com

      - name: Deploy to Cloud Run
        id: deploy
        uses: google-github-actions/deploy-cloudrun@v0
        with:
          service: ${{ env.SERVICE }}
          region: ${{ env.REGION }}
          # NOTE: If required, update to the appropriate source folder
          source: ./
      
      - name: Download SQL Auth Proxy
        run: "wget https://dl.google.com/cloudsql/cloud_sql_proxy.linux.amd64 -O cloud_sql_proxy"
        
      - name: Make the Cloud SQL Auth proxy executable
        run: "chmod +x cloud_sql_proxy"
        
      - name: Start the Cloud SQL proxy
        run: "./cloud_sql_proxy -instances=direct-album-348214:us-central1:nest-fun-basic=tcp:5432 & npx prisma migrate deploy"      

      # If required, use the Cloud Run url output in later steps
      - name: Show Output
        run: echo ${{ steps.deploy.outputs.url }}
```

The migrations worked for me in the Cloud SQL instance!

Would be great if "running Prisma migrations against Cloud Run" was added to the Prisma docs - Cloud Run is becoming more and more popular - I won't be the last person to run into the above issue!

**Reply by janpio** [reply]
Thanks, that sounds like something we might want to add to a docs article...

**Reply by tomwray13** [reply]
@janpio Hey Jan! If I have time I'll open a PR and write the docs article for this 😃

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
